From ecf46292bee81de901e102ca9b594db0ec04ddaf Mon Sep 17 00:00:00 2001
From: Bai Ping <ping.bai@nxp.com>
Date: Tue, 13 Dec 2016 16:57:58 +0800
Subject: [PATCH 6/7] MLK-13601-02 ARM: imx: Add fuse check support for imx6ull

On i.MX6ULL, some part can run at 800MHz setpoint.
we need to use the speed grading fuse to disable the unsupported
setpoint. speed grading fuse define as below:
    2'b00: Reserved;
    2'b01: 528MHz;
    2'b10: 792MHz;
    2b'11: Reserved;

Signed-off-by: Bai Ping <ping.bai@nxp.com>
(cherry picked from commit e8b9e8acb5ee44f48e048d744ccae4b6b02ef6a6)
---
 arch/arm/mach-imx/mach-imx6ul.c | 25 ++++++++++++++++++-------
 1 file changed, 18 insertions(+), 7 deletions(-)

diff --git a/arch/arm/mach-imx/mach-imx6ul.c b/arch/arm/mach-imx/mach-imx6ul.c
index 034db69..f20a9fa 100644
--- a/arch/arm/mach-imx/mach-imx6ul.c
+++ b/arch/arm/mach-imx/mach-imx6ul.c
@@ -66,7 +66,11 @@ static void __init imx6ul_opp_check_speed_grading(struct device *cpu_dev)
 	void __iomem *base;
 	u32 val;
 
-	np = of_find_compatible_node(NULL, NULL, "fsl,imx6ul-ocotp");
+	if (cpu_is_imx6ul())
+		np = of_find_compatible_node(NULL, NULL, "fsl,imx6ul-ocotp");
+	else
+		np = of_find_compatible_node(NULL, NULL, "fsl,imx6ull-ocotp");
+
 	if (!np) {
 		pr_warn("failed to find ocotp node\n");
 		return;
@@ -82,17 +86,25 @@ static void __init imx6ul_opp_check_speed_grading(struct device *cpu_dev)
 	 * Speed GRADING[1:0] defines the max speed of ARM:
 	 * 2b'00: Reserved;
 	 * 2b'01: 528000000Hz;
-	 * 2b'10: 700000000Hz;
+	 * 2b'10: 700000000Hz(i.MX6UL), 800000000Hz(i.MX6ULL);
 	 * 2b'11: Reserved;
 	 * We need to set the max speed of ARM according to fuse map.
 	 */
 	val = readl_relaxed(base + OCOTP_CFG3);
 	val >>= OCOTP_CFG3_SPEED_SHIFT;
 	val &= 0x3;
+	if (cpu_is_imx6ul()) {
+		if (val < OCOTP_CFG3_SPEED_696MHZ) {
+			if (dev_pm_opp_disable(cpu_dev, 696000000))
+				pr_warn("Failed to disable 696MHz OPP\n");
+		}
+	}
 
-	if (val != OCOTP_CFG3_SPEED_696MHZ) {
-		if (dev_pm_opp_disable(cpu_dev, 696000000))
-			pr_warn("Failed to disable 696MHz OPP\n");
+	if (cpu_is_imx6ull()) {
+		if (val != OCOTP_CFG3_SPEED_696MHZ) {
+			if (dev_pm_opp_disable(cpu_dev, 792000000))
+				pr_warn("Failed to disable 792MHz OPP\n");
+		}
 	}
 	iounmap(base);
 
@@ -161,8 +173,7 @@ static void __init imx6ul_init_irq(void)
 static void __init imx6ul_init_late(void)
 {
 	if (IS_ENABLED(CONFIG_ARM_IMX6Q_CPUFREQ)) {
-		if (cpu_is_imx6ul())
-			imx6ul_opp_init();
+		imx6ul_opp_init();
 		platform_device_register_simple("imx6q-cpufreq", -1, NULL, 0);
 	}
 
-- 
1.9.1

